### ISSUE TITLE
Allow correlating writes with feed notifications

### ISSUE THREAD

**Issue by coffeemug** [rethinkdb/rethinkdb#4629]
I talked to @Slava today about missing RethinkDB features that would make a Meteor integration much simpler. He said he only needs #3579, and ability to correlate writes with feed notifications. This issue is for the latter feature.

Imagine you have a chat application; when the user sends a message you want to show it on the client right away (possibly greyed out) to compensate for latency (which results in a much better user experience). When the notification later shows up in the feed, you then want to update the color to indicate that the message has been broadcast to everyone. In order to do that well, we need monotonically increasing write ids on per-connection basis.

Here is an example:

```js
> r.table('foo').insert({name: 'Slava'}).run()
{
    inserted: 1,
    write_id: [conn_id, timestamp],
    // ...
}
```

Then on the feed:

```js
> r.table('foo').changes()
{
    new_val: ...,
    old_val: ...,
    write_id: [conn_id, timestamp]
}
```

This way the client can correlate the write with the feed notification and do latency compensation.

There are a few outstanding questions:

- How do we return write ids when the user does a batch write?
- Does it always make sense to return it on a feed (e.g. what about complex feeds like `orderBy.limit`, etc.)? Will there ever be a case when we can't return a write id on the feed?

Also, ideally the timestamp would be user readable -- e.g. `time-since-epoch_integer` (the `integer` is there to dedup two writes with the same timestamp).

Putting in 2.2.

/cc @mlucy @danielmewes

**Comment by timmaxw** [comment]
Instead of `[conn_id, timestamp]` it might be better to use `[branch_id, counter]`. We already internally assign one of those to every write.

Another advantage is that it might allow some kind of eventual integration with resuming changefeeds using B-tree timestamps. For example, we could put a single set of tokens on the changefeed and use them both for correlating writes and for resuming the feed.

**Comment by mlucy** [comment]
The advantage of `[branch_id, counter]` would be that for squashing changefeeds where 5 changes to the same document are squashed together we could just take the latest stamp and count that as the stamp for the squashed change.  The disadvantage of that is that the client would have to know how the branches relate to each other in order to figure out whether `[A, 1]` is before or after `[B, 1]` (e.g. they issue a write on a document and get back a stamp `[A, 1]`, then they receive a change for that document with stamp `[B, 1]`, and they don't know whether that squashed change includes their write or not unless they know how branch `A` relates to branch `B`).

I think maybe the way we should do this is a map from machine IDs to timestamps.  The stamp for a single unsquashed change would just be the machine ID and the timestamp, the stamp for a squashed change that includes two changes from the same machine would be that machine's ID and the latest timestamp, and the stamp for a squashed change that includes two changes from two different machines would be a map with two entries.  (I think machine IDs are probably better than connection IDs.)

**Comment by mlucy** [comment]
We could also just not worry about squashing changefeeds for the first version, but I vaguely recall that Meteor wanted that feature.

**Comment by danielmewes** [comment]
Independently of relating different branches for the same region to each other, `[branch_id, counter]` has the general problem of not being monotonically increasing per connection. If different writes hit different shards, they will have different `branch_id`s with no useful relation to each other.

@slava what are the guarantees that you actually need with respect to the IDs being monotonically increasing? Does this have to hold on a per-primary-key basis or on a per-connection basis? E.g. if you write to two different documents A and B, is it important that you can figure out from the returned IDs which one was sent first? Or is it enough to have that guarantee if you submit two writes to the same document?

**Comment by mlucy** [comment]
I was imagining that monotonically increasing IDs were important for squashing, and more generally so that if machine A sends `n` writes and tells machine B to wait for them it doesn't have to send O(n) stamps to `B` (monotonically increasing per-machine IDs work for this because you just have to send O(machine) stamps to `B` and your cluster probably isn't that big).

**Comment by danielmewes** [comment]
> How do we return write ids when the user does a batch write?

If we use `[client_id, counter]`, we could use the same id for all writes in the batched write.
We could assign the counter on a per-query basis on the parsing node. All `write_t`s generated from that query will have the `client_id`/`counter` pair, and the changefeeds will submit that with their notifications when the writes are executed.
One problem with this is that it's not clear when the whole query has completed.

**Comment by danielmewes** [comment]
Actually thinking about this more I'm still a bit fuzzy on the use case for this feature.

Specifically: The way I understand this, the idea is to know when a given write has been reflected in the database and has started to be broadcasted to all subscribed users.

In the original proposal the idea was to accomplish this by getting some sort of ID for a given write, and then waiting for that ID to show up on the changefeed. However in that case, one could just as well simply wait on the write to be acknowledged, which has about the same effect.

Is there more to this? @slava would you mind clarifying?

**Comment by mlucy** [comment]
> One problem with this is that it's not clear when the whole query has completed.

I think that sort of sinks the idea.  I think the primary use case here would be issues a batch of writes on `A`, sending a small amount of data to `B` (which has an open changefeed), and then being able to know when `B` has received the entire batch of writes.

**Comment by timmaxw** [comment]
Another option would be a command called `table.flush(arbitrary_token)`, which causes a special message containing `arbitrary_token` to be inserted into every changefeed on `table`, with the guarantee that the extra message would come after every write that had been completed before `r.flush()` started.

**Comment by coffeemug** [comment]
> E.g. if you write to two different documents A and B, is it important that you can figure out from the returned IDs which one was sent first?

From my conversation with Slava, if you send a write A and get a timestamp Ta, then send a write B and get a timestamp Tb, and then see Tb on a feed, we must guarantee that Ta should also have been reflected in the feed (which I think is equivalent to being able to figure out which one was sent first).

**Comment by danielmewes** [comment]
> Which I think is equivalent to being able to figure out which one was sent first.

That's possibly true if the client has only one outstanding write at a time. In general that's not true in the current implementation, because different shards will have their own ordering for writes that is independent of the order in which the writes were submitted by a client.

I imagine we could modify changefeeds to adhere to the client sequence ordering if this is really necessary.

**Comment by Slava** [comment]
> I was imagining that monotonically increasing IDs were important for squashing, and more generally so that if machine A sends n writes and tells machine B to wait for them it doesn't have to send O(n) stamps to B (monotonically increasing per-machine IDs work for this because you just have to send O(machine) stamps to B and your cluster probably isn't that big).

Yes, I think this is exactly my understanding. The monotonically increasing IDs help to see a squashed change and compare it to the ones application is waiting for.

> Specifically: The way I understand this, the idea is to know when a given write has been reflected in the database and has started to be broadcasted to all subscribed users.

> In the original proposal the idea was to accomplish this by getting some sort of ID for a given write, and then waiting for that ID to show up on the changefeed. However in that case, one could just as well simply wait on the write to be acknowledged, which has about the same effect.

This is not exactly what I've proposed to @coffeemug. We don't care about DB acknowledging the write (we get it already in the current driver). We care to see when a write appears in the changefeed observed by application.

**Comment by danielmewes** [comment]
Thanks @Slava for chiming in.

> Yes, I think this is exactly my understanding. The monotonically increasing IDs help to see a squashed change and compare it to the ones application is waiting for.

Ok that makes sense. That sounds like the IDs need to be monotonically increasing per primary key then, which suggests that we could indeed use the replication `counter` assigned by the shard's primary (possibly together with the branch ID).

**Comment by Slava** [comment]
I don't understand what "per primary key" means? Can you clarify?

**Comment by danielmewes** [comment]
@Slava What I meant by that was that the assigned "write IDs" are monotonically increasing gor any given document in the table, but not necessarily across documents.
Say for example we have a squashing changefeed on the `test` table: `r.table('test').changes({squash: true})`, and the table has two documents `{id: "a", val: 1}, {id: "b", val: 2}`. Now the application does two writes that update document `"a"`. Let's say those get the write IDs `a100` and `a105` respectively. Now once you see a change with ID `a105`, you will know that all writes to `"a"` with smaller write IDs will be reflected in that, in this example `a100` will be in there as well.
However if you see a change `b200`, that does not mean that `a100` and `a105` have already been reflected on the changefeed. (However the writes `b100`, ..., `b199` would all be guaranteed to be reflected.)
I hope that makes sense. Basically the ordering of IDs would only be guaranteed for an individual document, but not across the whole table.


That being said, the primary key criteria might not work too well with squashing `orderBy.limit` changefeeds, because I believe those can squash changes involving multiple different primary keys. So in that case the ID for a squashed change wouldn't be in a total ordering with respect to a set of writes. It might be necessary to not use squashing in that case.

**Comment by Slava** [comment]
@danielmewes I see.

I think if the application-level driver gets the write IDs per write, it should be enough to have "per primary key" monotonically increasing write IDs.

But you need to think more, what happens when a query touches a lot of documents? Would you return write IDs for all of them? Like a query of `UPDATE collection SET field = field + 1 WHERE field > 20` and it updates 200 rows?

If the write IDs are monotonically increasing globally, then you can just return the latest ID.

**Comment by danielmewes** [comment]
Yeah I'm not sure, but I think it's pretty hard to implement globally monotonically increasing IDs, even if we can do so on a per-client basis.

Something like this could be an option:
(details for the implementation follow, you can skip this part if you're not interested in those)

- On the parsing node (the server the client is connected to), assign a client ID to each connected client and generate IDs `[client_id, table, counter]` locally.
- Send the `[client_id, table, counter]` pair with each write, and make it be distributed from the changefeed server to the changefeed clients as well (the changefeed client in this case is a component inside the RethinkDB server, which lives on the server that a (ReQL) client with an open changefeed is connected to)
- For a given squashed result on its queue, the changefeed client keeps track of the highest counter value it has seen, but only for writes whose `client_id` matches the client that's running the changefeed.
- The changefeed client also uses a `fifo_enforcer_sink_t` (or something similar) to re-establish the ordering of write notifications according to their counter values. This is important because notifications from different shards can arrive in different orders that do not necessarily correspond to their counter values.
- When a squashed result is sent to the ReQL client, the highest counter value is sent with the result. If no counter has been seen from the `client_id` associated with the changefeed (i.e. all writes were done by other clients), the changefeed sends a `null` write ID or something like that.

This would establish monotonically increasing IDs *per table* and per client.
That is: if you get a change notification with a given ID, you know that all changes with smaller IDs submitted by the current client will have been reflected on the changefeed (even if they were for different documents).

**Comment by coffeemug** [comment]
Why do we need `table` in there? Wouldn't be just as easy to do `[client_id, counter]` (which I think is strictly better for users)?

**Comment by danielmewes** [comment]
@coffeemug We cannot increment the counter across tables, or the changefeed client would have no way of reconstructing the order of writes.

A given changefeed client only received changes for one table. Suppose a client performs the following writes:
    `["client_X", 1]` to table "a", `["client_X", 2]` to table "b", `["client_X", 3]` to table "a"
Now the changefeed client that listens for changes on table "a" will receive a notification with ID `["client_X", 1]` first, followed by `["client_X", 3]`. When it receives that second notification, it has no way of knowing whether it has to hold back that notification because it needs to wait on `["client_X", 2]`, or not.

With the table component, the IDs would look like this instead:
    `["client_X", "a", 1]`, `["client_X", "b", 1]`, `["client_X", "a", 2]`

This allows the changefeed client to know when a change to its table (for its client) is missing.

**Comment by danielmewes** [comment]
It should also be noted that implementing this will limit our options for future optimizations in changefeeds. It relies on the changefeed client to always receive all notifications for the whole table, even if all changefeeds are just on a specific key range (`between`) or for a certain `filter`.

It's probably easier from our side to assign individual IDs to every modified document. I can see though that this will become a limiting factor in some types of queries (e.g. a `r.table(...).delete()`) can easily touch millions of document).

**Comment by coffeemug** [comment]
> A given changefeed client only received changes for one table.

Note that this won't be true for `eq_join` feeds, and possibly other feeds in the future.

> Suppose a client performs the following writes:
    `["client_X", 1]` to table "a", `["client_X", 2]` to table "b", `["client_X", 3]` to table "a"
Now the changefeed client that listens for changes on table "a" will receive a notification with ID `["client_X", 1]` first, followed by `["client_X", 3]`. When it receives that second notification, it has no way of knowing whether it has to hold back that notification because it needs to wait on `["client_X", 2]`, or not.

I think it's ok for the client that listens on table `a` to receive `["client_X", 1]` first, followed by `["client_X", 3]`. I think we shouldn't make any guarantees about the timestamp other than the fact that it's increasing, and that it can be correlated back to the write. I think it's ok to skip ids; we may even want to use actual unix time since epoch. The idea is that when you receive a notification with a timestamp `X`, you know that all writes with a timestamp less than or equal to `X` have been reflected in the feed; giving more information seems limiting and unnecessary IMO.

**Comment by danielmewes** [comment]
>  Note that this won't be true for `eq_join` feeds, and possibly other feeds in the future.

True. I don't know how this will work out then. It won't get easier for sure :-(

I think I might have been unclear. By the "changefeed client", I meant the thing that lives on the parsing node, not the ReQL client. Skipping IDs there is not a choice if we want to maintain the invariant that once a given ID is received on a changefeed, all writes with smaller IDs will have been reflected in it. If we skip IDs, we cannot recover the ordering of writes to different shards as far as I can tell, so there's no chance we can maintain that invariant.

Note that all of this only applies to @Slava's suggestion here:
> But you need to think more, what happens when a query touches a lot of documents? Would you > return write IDs for all of them? Like a query of `UPDATE collection SET field = field + 1 WHERE field > 20` and it updates 200 rows?
> 
> If the write IDs are monotonically increasing globally, then you can just return the latest ID.

If we're ok with generating one write ID for each affected primary key, then we won't have that problem and the implementation will be a lot simpler. We can for example just use replication timestamps in that case, and the changefeed client (the thing on the server) doesn't need to care about reconstructing ordering across documents.

**Comment by coffeemug** [comment]
> If we're ok with generating one write ID for each affected primary key

Why wouldn't we be ok with that?

**Comment by danielmewes** [comment]
@coffeemug Because if you do large writes e.g. `r.table(...).delete()` or the `UPDATE collection SET field = field + 1 WHERE field > 20` that @Slava mentioned, we will potentially have to return a huge number of write IDs to the write query. This can go into the gigabytes.

I feel like this might be an acceptable restriction assuming it makes our implementation a lot easier, but it's not perfect.

**Comment by mlucy** [comment]
Here's an alternate proposal:

* Timestamps are a set of <machine ID, wall time> pairs.
* When a batch write is first issued, we shard it and send it to multiple machines.
  - On each primary, for each write in the batch write we assigned the write a stamp of `{primary: current_time}`.  This is the stamp that is sent with that write to the changefeed.  We return the stamp `{primary: max(current_time)}` (so the latest stamp from all the writes in the batch).
  - During the unsharding phase, we simply union all these sets together (so if we have `{m1: t1}`, `{m1: t2}`, and `{m2: t3}` unsharding produces `{m1: max(t1, t2), m2: t3}`).
* Changefeeds maintain a set of `<machine ID, latest wall time>` pairs and send a stamp at the end of every batch as a `state` message.
* Clients just need to watch for these state messages, and if they see a state message with a stamp that strictly dominates the stamp they received for a write (the stamp is >= the stamp they received for every machine) they know that write is reflected in the changefeed.  (In the case where they have an unsharded table, they just need to compare two numbers during this step.)
  - Optionally we provide a way to automate this in the clients so that people don't have to write the comparison function themselves.

I think that solves all the major problems, and has the advantage that it also works for coordinating writes with changefeeds opened by other clients (I think it's good to be general here if we can).

One long-term complication is failover.  Long-term we'd like changefeeds to survive failover, and I'm not 100% sure how this can be adapted to support that.  (Although it would probably be fine to have an additional sync message or something in that case that the clients have to handle.)

**Comment by danielmewes** [comment]
@mlucy Assuming we can make that work, I think that doesn't sound too bad. Especially if we implement the comparison function in the clients (I think the write ID / timestamp sets should probably be an opaque type in that case, with only a comparison function).

I have a few technical concerns thought. I'm not sure how easy it is to solve these:

- we still need to make sure that all previous writes from a given machine have reached the changefeed before we send the `state` message. That will require knowledge on each machine about which timestamps have been assigned on the machine, and it will need a way to tell the changefeed::client_t when all writes previous to a given timestamp have been transmitted to it. One way I can think of for this is using a single fifo source per machine, and sending the resulting tokens with each change. This isn't ideal because it involves cross-thread communication.
- wall time can generate duplicates, so that's insufficient. If we introduce a per-machine fifo source to solve the previous problem, we could just use the IDs it generates instead.

I wonder if it could work to replace `[machine, wall time]` by `[branch_id, counter]` in your proposal. So during unsharding, we would merge multiple counters for the same `branch_id` into `[branch_id, max(counter)]`. With each state message, we would send one counter per branch which isn't really compact, but might be ok.

@mlucy: Do changefeeds already guarantee the ordering of events per shard, even across different primary keys within the shard?

**Comment by mlucy** [comment]
You're right, `<shard, timestamp>` pairs are what we need to actually provide the ordering guarantee (or, even better, `<changefeed_server_uuid, timestamp>`).  And yeah, changefeeds already guarantee per-shard ordering.

Is it actually true that we can't generate duplicate-free wall times?  I guess if these things are opaque then a counter instead if fine.  We can even re-use the counters that are already on `changefeed::server_t`.

---

I don't really know enough to say whether we could use `branch_id, counter`.  That certainly seems cleaner, and we should only generate a new branch ID when resharding or failover occurs, which currently causes us to disconnect anyway.

**Comment by mlucy** [comment]
Nevermind, I guess you can always get duplicate wall times if you change the system clock.

**Comment by danielmewes** [comment]
That and we might request the wall time on multiple threads in the same moment.

I think `[changefeed_server_uuid, changefeed_counter]` would currently be equivalent to `[branch_id, replication_counter]`. So as far as I can tell either should work, and they are indistinguishable from the client's perspective so we could swap them out later if we want to.
New branch IDs are also generated when a primary is restarted, even if no failover occurs in between.

**Comment by mlucy** [comment]
OK, let's do it with the `changefeed_server_uuid, changefeed_counter` pair because those are already sent with every change.  This should actually be pretty straightforward to implement.

**Comment by danielmewes** [comment]
Sounds good to me! :-)

**Comment by mlucy** [comment]
I didn't realize this was scheduled for 2.2, but I still agree with what we decided before (returning stamps which are maps from changefeed server UUIDs to changefeed stamps).

**Comment by danielmewes** [comment]
:+1: Same here.

**Comment by danielmewes** [comment]
We seem to agree on the form of the timestamps, i.e. `[changefeed_server_uuid, changefeed_counter]`.

For multi-document writes, we will merge the timestamps in the `write_id` field by keeping only the highest `changefeed_counter` for any given `changefeed_server_uuid`.
In general `write_id` is going to be an array of such timestamp pairs.

There are (at least) four open questions:
* Are we going to send the `write_id` with every changefeed notification, or as a separate `state` document? Putting them into the changes themselves might considerably increase the size of squashed changefeed notifications, so using a separate `state` might be better.
* What should the field be called? I think `write_id` is a bad name because it doesn't convey the fact that there's a partial ordering on the IDs. I think `write_timestamp` might be a better name?
* Should we require timestamp information in the changefeeds to be enabled explicitly or do we always want them to be there? I think at least if they go in their own `state` documents, they should be optional.
* Should we provide a comparison function for timestamp arrays in the clients? If yes, how should it be called? Maybe the timestamp array itself should be a ReQL pseudo-type, that gets deserialized to a custom class instance with a method for comparing them?

**Comment by danielmewes** [comment]
The last time I looked at this I missed that there were a bunch of open details about it unfortunately.

We might have to move this out of the 2.2 milestone and into 2.3 to keep the 2.2 release within the planned time frame. I'll think about that a bit.

**Comment by danielmewes** [comment]
Actually even if we don't make the timestamps their own pseudo type, I think they should probably be an object of the form
```
{
  "changefeed server 1 UUID": changefeed_counter1,
  "changefeed server 2 UUID": changefeed_counter2,
  ...
}
```

That would make the check for whether a given write timestamp has been reflected in a changefeed more convenient to implement.

**Comment by mlucy** [comment]
> In general write_id is going to be an array of such timestamp pairs.

I think the map notation you mention in your later post is better.

> Are we going to send the write_id with every changefeed notification, or as a separate state document? 

I think either way is fine, honestly.  Sending it with every notification won't make them *that* much bigger.  I guess I think a state document is slightly better, but I don't have a strong opinion.

> What should the field be called?

`write_timestamp` or `write_timestamps` seem find to me.

> Should we require timestamp information in the changefeeds to be enabled explicitly or do we always want them to be there?

I think you should always be able to turn them on and off.  I think we should probably leave them off by default no matter where we put them, just to make the interface less intimidating for new users (since they are for a fairly advanced use case).

> Should we provide a comparison function for timestamp arrays in the clients?

Probably!

> Maybe the timestamp array itself should be a ReQL pseudo-type, that gets deserialized to a custom class instance with a method for comparing them?

That solution is a little complicated, but it's probably more user friendly than having an `r.timestamp_lt` function or something.

We could also offer utility functions in the asynchronous interfaces like "run this code when this timestamp is reached", which I think would make a lot of code simpler.

---

> We might have to move this out of the 2.2 milestone and into 2.3 to keep the 2.2 release within the planned time frame.

I'm definitely a little behind on 2.2 since the unsharding work is taking longer than I thought it would, so that would be another reason to consider bumping it.

**Comment by coffeemug** [comment]
I don't think we should bump it out of 2.2 for story-telling reasons. This feature fits into a coherent story where we make it much easier for framework developers to do what they need on top of RethinkDB; without it the story for the release is much, much weaker.

I think it would be much better to delay 2.2 a little bit, include this into 2.2, and start some of the 2.3 work in parallel.

**Comment by coffeemug** [comment]
>* Are we going to send the `write_id` with every changefeed notification, or as a separate `state` document?
* Should we require timestamp information in the changefeeds to be enabled explicitly or do we always want them to be there? I think at least if they go in their own `state` documents, they should be optional.

I think we should appropriate the `includeStates` optarg for this. When it's false, the user would never see the timestamps; when it's true, it would be a part of the `state` document (e.g.: `{state: 'timestamp', write_timestamp: foo}`).

This seems much cleaner to me, and wouldn't require changing the interface very much.

> * What should the field be called? I think `write_timestamp` might be a better name?

:+1: for `write_timestamp`

> Should we provide a comparison function for timestamp arrays in the clients? If yes, how should it be called? Maybe the timestamp array itself should be a ReQL pseudo-type, that gets deserialized to a custom class instance with a method for comparing them?

I think since it's a fairly advanced feature, for the first release I'd document how the timestamps are to be compared and let the users do it themselves. If there is a lot of demand for baked in comparison code, we can add it later, but I don't think it's necessary for v1 of this.

**Comment by danielmewes** [comment]
@coffeemug Yeah I thought so. It certainly fits well into 2.2.

(@mlucy)
> I think you should always be able to turn them on and off. I think we should probably leave them off by default no matter where we put them, just to make the interface less intimidating for new users

(@coffeemug)
> I think we should appropriate the `includeStates` optarg for this.

:+1:

> I think since it's a fairly advanced feature, for the first release I'd document how the timestamps are to be compared and let the users do it themselves.

That's probably right. I think we can get away with that for now. That also avoids us having to implement the logic in the Python and Ruby drivers, which wouldn't help Meteor at all.

**Comment by danielmewes** [comment]
Marking settled as described with a `write_timestamp` field in write responses, adding an option to `includeStates` to explicitly enable the feature on changefeeds, and having the information supplied as `state` documents on the feed.

**Comment by opsb** [comment]
Just adding a :+1: for including the timestamps in separate state document. If a client is subscribed to a query that doesn't include all results from a table then it's possible they could write a change to the table which wouldn't be included in their results. If separate state documents are used then this isn't a problem, the state document can be included in the changefeed even though the write itself won't be.

**Comment by danielmewes** [comment]
@opsb Good point!

**Comment by faceyspacey** [comment]
is it just me, or is RethinkDB a match made in heaven for Meteor (and vice versa)? If that's the case, this is the one thing holding back Slava's Meteor-RethinkDB integration. Could be big for both camps.

**Comment by coffeemug** [comment]
Yep. We're on this.

**Comment by opsb** [comment]
I'm also hanging out for this one :) Need it for integration with [orbit.js](https://github.com/orbitjs/orbit.js).

**Comment by muloka** [comment]
:+1:

**Comment by danielmewes** [comment]
A quick update: We're very close to finalizing things for the 2.2.0 release and start with final pre-release testing, and this just barely missed the deadline. It should be in the next release very soon though.

**Comment by opsb** [comment]
Dang :) If it's going to be in master soonish that will be good enough for me though (the orbit integration will probably take a little while to put together anyway so it's unlikely to be going into production before the next release). Do you guys have instructions for building and running on OS X?

**Comment by danielmewes** [comment]
@opsb It's basically what's described here http://rethinkdb.com/docs/build/ under "Building from git". On OS X you should install XCode first. I think that will cover most (if not all) essential dependencies. 
@larkost might have some additional notes.

**Comment by vlasky** [comment]
@danielmewes How is this coming along? You probably know that the Meteor community is itching to embrace RethinkDB.

**Comment by coffeemug** [comment]
This should be in the next release (2.3).

**Comment by opsb** [comment]
That's great news @coffeemug! Looking forward to getting this integrated.

**Comment by opsb** [comment]
@coffeemug I'm not sure how far along this is, has the api been decided yet?

**Comment by danielmewes** [comment]
@opsb We have an API and a partial implementation. We're trying to ship this for non-squashing changefeeds with the 2.3.0 release.

@mlucy could you briefly describe the API we implemented? I forgot the details of it.

**Comment by opsb** [comment]
I had a quick scan of #next but couldn't tell whether this is in yet, has it been merged in or is it still wip?

**Comment by danielmewes** [comment]
@opsb and others:
Unfortunately I have bad news. We have a partially working implementation in the branch https://github.com/rethinkdb/rethinkdb/tree/mlucy_write_stamps , but we haven't been able to properly test it yet. Due to all the other things ( https://github.com/rethinkdb/rethinkdb/milestones/2.3 ) that still need to happen for the 2.3 release, I decided to move this out of the 2.3 milestone and instead ship the write stamps with RethinkDB 2.4. We could theoretically merge what we have right now, but there's a very high risk that it will have bugs and I wouldn't feel good about doing so.

Again apologies for the delay. I know you have been waiting for this for a while.

**Comment by opsb** [comment]
Gah, disappointing :( Completely understand about not wanting to merge something shaky though, I certainly appreciate the general reliability of rethinkdb. 

I'm keen to at least get what's there running locally so I can start to develop against it. My c++ isn't so hot these days (must have been nearly 15 years since I last wrote any), if you guys could provide any pointers regarding how to use the timestamps in the client it would be greatly appreciated :)

**Comment by danielmewes** [comment]
@opsb We can discuss the details of how to use the timestamps on Slack. My user ID is danielmewes on https://rethinkdb-team.slack.com . If you aren't registered yet, you can go to http://slack.rethinkdb.com to get signed up.

**Comment by opsb** [comment]
Just wondering, if the current implementation doesn't support squashing, does this mean that it's possible to match received changed to writes on a one to one basis?

**Comment by danielmewes** [comment]
@opsb I think it is, at least as far as point writes are concerned (writes that change only one document). For writes that affect multiple documents, you can see a variable number of timestamps, some of which will be shared across multiple document changes.

**Comment by opsb** [comment]
That's really useful for scenario that I'm looking at :) A further clarification, if a timestamp can be shared across multiple document changes, would they only ever be shared within a single write or could they be shared across multiple concurrent writes?

**Comment by danielmewes** [comment]
@opsb Only between writes done as part of the same query.

**Comment by makkesk8** [comment]
Its been a year, any updates?

**Comment by danielmewes** [comment]
@makkesk8 The status is still what's written in this comment: https://github.com/rethinkdb/rethinkdb/issues/4629#issuecomment-203610958

Basically: We have an implementation which needs some final touches, and will try very hard to ship it in RethinkDB 2.4.

**Comment by makkesk8** [comment]
@danielmewes Alright, thanks.

**Comment by alawi** [comment]
Hello folks - coming from the Meteor camp this seems like a much needed feature to have a working integration. Going through the comments it seems that the implementation is almost complete but it requires more testing. I personally think this feature should get a higher priority since it's a clear win-win situation. Meteor has a considerable community size and this will definitely gives RethinkDB more visibility (that is how I came to know it)  and on the other hand Meteor seems to have scaling issues that RethinkDB should be able to solve. There is a bit of churn in the Meteor community and I think this is the right time to get this moving and if there is anyway I can help I'd love to.

**Comment by gustojs** [comment]
Sorry for pinging that, but I saw that the list of commits for 2.4 was closed already, so I wanted to ask if this made it to the list or was it pushed back again to 2.5.

Quoting MDG member comment from 25 days ago - "If that's resolved, I think it will be easy to integrate Rethink"

**Comment by AtnNn** [comment]
This issue hasn't made the list. I believe the implementation is still incomplete.

**Comment by Extarys** [comment]
Sad, some guys still wait on this to include it in meteor
